docs(MADR): spiffe compliance - context and user stories (#13699)
kumahq/kuma: docs/madr/decisions/073-spiffe-compliance.md

In SPIFFE, a Trust Domain is a logical boundary for identity and trust, similar to a DNS domain. Services within the same trust domain can securely identify and trust each other. Communication across different trust domains (such as between organizations or clusters) requires federation, which establishes trust relationships so that identities from one trust domain are recognized and trusted by another.

Kuma currently uses the mesh name as the trust domain. While this works inside Kuma, it doesn’t clearly separate trust boundaries, and it limits how easily Kuma can connect with external services or other applications in the same trust domain.

This setup causes problems when:

* You want to connect Kuma with other SPIFFE-compliant systems outside the mesh
* You run multiple clusters that need to trust each other

> * You work in environments that require strict identity standards for security or compliance reasons

Reviewer: missing:

- You have identity not coming from Kuma (.e.g SPIRE)

To fully support SPIFFE and modern security practices, Kuma should:

* Issue certificates that include only one SPIFFE-compliant URI SAN (currently we generate `spiffe://<mesh>/<service>` for each inbound and `kuma://kuma.io/<tag_name>/<tag_value>` for each tag)
* Let users set the trust domain explicitly, not tie it to the mesh name

> * Allow using other systems (like SPIRE) to manage identities

Reviewer: I think one very important thing:

- Kuma should separate issuing identity from validating trust.

### As a user, I want to be able to rotate or change the Certificate Authority (CA) used for issuing mTLS identities without interrupting existing connections or causing service downtime.

During the entire process, communication between services must remain secure to ensure uninterrupted, encrypted traffic. This is essential for maintaining zero-trust security guarantees and ensuring operational reliability in production environments.

### As a user, it should not become harder to enable or manage mTLS in either single-zone or multi-zone deployments than it is today.

Currently, users can enable mTLS in both single-zone and multi-zone deployments just by configuring it in the Mesh resource. As we introduce features like trust domains, external identities, or federation, we must preserve this simplicity and avoid adding unnecessary complexity to the mTLS setup process.

> ### As a user, I want to be able to communicate with applications in the different trust domain.

Reviewer: ```suggestion
### As a user, I want to be able to communicate with applications in different trust domains.
```

By assigning unique trust domains per cluster, we achieve security isolation: if one cluster is compromised, the others remain secure. Additionally, it helps with identity scoping, since we can clearly determine the origin of each identity. Another important aspect is enabling rotation or migration. During the migration, we could use two different trust domains to facilitate the transition from one identity provider to another.

Default trust domain: `spiffe://<mesh>.<clusterId>.kuma.io/`

### As a Mesh operator, I want to be able to migrate some workloads to another trust domain without interrupting traffic.

I have a workloads in a trust domain `example.com` and I want to move them to `example.org`. I should be able to do it and don't break the traffic.

### As a user, I want certificate to be valid for a short period of time

Based on the [specification](https://spiffe.io/docs/latest/spiffe-about/spiffe-concepts/#spiffe-workload-api) certificates should be short-lived.

### As a user, I want to migrate from Kuma non-SPIFFE complient certificates to SPIRE

> We should allow users to migrate from Kuma issued certificates, which are not SPIFFE-compliant, to using SPIRE. This may require an intermediate step to first migrate to SPIFFE-compliant certificates, but the details will be specified later.

Reviewer: So that's a future story right?

> ## Future User Stories (things that are going to be implemented in future releases)

Reviewer: I think one is automatic CA rotation.

It's fairly straightforward:
```
Preparation Phase:

Generate a new root CA certificate and private key
The new root CA is prepared but not yet active
Both old and new root CAs exist simultaneously during the transition

Bundle Update:

The SPIFFE trust bundle is updated to include both the old and new root CA certificates
This bundle is distributed to all workloads through the Workload API
Workloads now trust certificates issued by either CA

Gradual Transition:

New X.509-SVIDs (SPIFFE Verifiable Identity Documents) are issued by the new CA
Existing SVIDs continue to be valid until they expire
Services can verify certificates from both CAs during this period

Completion:

Once all SVIDs have been renewed with the new CA, the old root CA is removed from the trust bundle
Only the new CA remains active
```

However, we shouldn't do this just right now.

### As a user, I want to have a SPIFFE-compliant SAN in a non-Kubernetes environment

It would be beneficial to support connectivity between different environments. Unfortunately, our services currently lack proper identity on Universal, which makes this challenging.

### As a user, I want to federate different trust domains using the Federation API exposed by the control plane

> It should be possible to federate different trust domains without creating additional resources, only by using the federation endpoint exposed by the control plane. This would be a valuable feature, enabling interoperability between different SPIFFE-compliant systems.

Reviewer: The federation endpoint is part of SPIRE not spiffe no?

Author: https://github.com/spiffe/spiffe/blob/main/standards/SPIFFE_Federation.md#2-introduction

### As a user, I want to use an intermediate CA instead of a root CA on the zone

When a user provides a root CA on the global control plane, we could generate a separate intermediate CA for each connecting zone. This would offer a stronger security model, ensuring that if a certificate is compromised in one zone, it does not impact other zones.

## Out of scope

### As a user, I want to specify certificates for a specific outbound

You should use MeshExternalService in this case to configure the outbound connection and provide the required certificates.

### As a user, I want my dataplanes to communicate with Istio services without going through a gateway

In this scenario, you should also use MeshExternalService and include the necessary certificates in the configuration to enable direct mTLS communication.

### As a user, I want to use two different sources of CA (Kuma and SPIRE) at the same time

This is a complex use case and can make the configuration error prone and difficult to manage.
We want to add some limitation to avoid issues:
* A dataplane's identity must come from a single source.

> * For trust domains, identities must be issued by one system only — either SPIRE or Kuma, never both for the same dataplane.

Reviewer: We're mixing things here I think.
I think we're getting the info (identity and trust) either from Spire or Kuma never both.
However, it's possible to add the trust coming from Spire inside Kuma and to implement SPIRE federation api in Kuma to retrieve Trust.

Author: Rephrased

* It's valid for one dataplane to get its identity from Kuma and another from SPIRE — communication between such dataplanes should still work.

To enable this setup, we’ll need to establish a federation mechanism — potentially by having Kuma expose a federation endpoint.

### As a user, I want to have trust domain per namespace

> I think we shouldn't allow higher granularity than zone or mesh.

Reviewer: I don't understand this limitation? AFAIU it shouldn't be complicated to just run 1 dataplane on a different identity.

### As a user, I want to provide MeshTrafficPermission for SPIFFEID

This should be covered by a separate MADR in a separate [issue](https://github.com/kumahq/kuma/issues/12374).

> 

Reviewer: I think these stories are worth adding:

- Changing trustDomain without changing CA. Or can we do it?

Reviewer 2: From SVID [docs](https://github.com/spiffe/spiffe/blob/main/standards/X509-SVID.md?plain=1#L54) on the CA certificate:

> A signing certificate SHOULD itself be an SVID. If present, the SPIFFE ID of a signing certificate MUST NOT have a path component, and SHOULD reside in the trust domain of any leaf SVIDs it issues. A signing certificate MAY be used to issue further signing certificates in the same or different trust domains.

So if trustDomain is part of the CA cert then I believe we can't change trust domain without changing the CA? But I'm a bit confused by their wording `If present`. So it it up to us to include SPIFFE ID into the CA certificate?

Author: Seems so, we should create a CA with SPIFFEID without path part, but `If present` suggest it's optional

* Tools and rules for issuing, rotating, and checking these identities

Right now, Kuma does something similar by generating SPIFFE-like IDs like `spiffe://<mesh-name>/<service-name>`. But it also adds extra SANs (like `kuma://kuma.io/version/v1`) based on tags. These extra SANs make Kuma's certificates non-compliant with SPIFFE, meaning they can’t be used directly with SPIFFE-based systems like SPIRE, or other tools that expect full compliance.

In SPIFFE, a Trust Domain defines which services are allowed to trust each other. It’s like a boundary — services inside the same trust domain can safely communicate. To talk across trust domains (for example, between two companies or clusters), you need federation, which lets them trust each other's identities.

Kuma currently uses the mesh name as the trust domain. While this works inside Kuma, it doesn’t clearly separate trust boundaries, and it limits how easily Kuma can connect with external services or other meshes.

This setup causes problems when:

* You want to connect Kuma with other SPIFFE-compliant systems outside the mesh

> * You run multiple clusters or meshes that need to trust each other

Reviewer 3: You mean service meshes here, right?

* You work in environments that require strict identity standards for security or compliance reasons

To fully support SPIFFE and modern security practices, Kuma should:

> * Issue certificates that include only one SPIFFE-compliant URI SAN

Reviewer 3: Could we mention what is the current state, and what URIs we are currently including?

### As a Mesh operator, I should be prevented from configuring the same trust domain in multiple zones when not using SPIRE as the identity provider.

This point is a bit tricky for me, because on one hand, it makes sense to allow users to define a trust domain for an entire mesh (which may span multiple clusters). On the other hand, preventing users from using the same trust domain across different clusters improves security compliance.

By assigning unique trust domains per cluster, we achieve security isolation: if one cluster is compromised, the others remain secure. Additionally, it helps with identity scoping, since we can clearly determine the origin of each identity.

### As a Mesh operator, I want to be able to migrate some workloads to another trust domain without interrupting traffic.

I have a workloads in a trust domain `example.com` and I want to move them to `example.org`. I should be able to do it and don't break the traffic.

> 

Reviewer 3: ```suggestion
```

* Tools and rules for issuing, rotating, and checking these identities

Right now, Kuma does something similar by generating SPIFFE-like IDs like `spiffe://<mesh-name>/<service-name>`. But it also adds extra SANs (like `kuma://kuma.io/version/v1`) based on tags. These extra SANs make Kuma's certificates non-compliant with SPIFFE, meaning they can’t be used directly with SPIFFE-based systems like SPIRE, or other tools that expect full compliance.

In SPIFFE, a Trust Domain defines which services are allowed to trust each other. It’s like a boundary — services inside the same trust domain can safely communicate. To talk across trust domains (for example, between two companies or clusters), you need federation, which lets them trust each other's identities.

Kuma currently uses the mesh name as the trust domain. While this works inside Kuma, it doesn’t clearly separate trust boundaries, and it limits how easily Kuma can connect with external services or other applications in the same trust domain.

This setup causes problems when:

* You want to connect Kuma with other SPIFFE-compliant systems outside the mesh

> * You run multiple clusters that need to trust each other

Reviewer 4: I'm not sure I 100% understand this - currently if you run multiple clusters (zones) and they share the same `mesh` then everything is fine right?

Author: Yes, but all have the same trust domain which might not be correct since each cluster can have a different one

Reviewer: So what you want is to have "zones" that have different trustDomains.

### As a user, I want to be able to rotate or change the Certificate Authority (CA) used for issuing mTLS identities without interrupting existing connections or causing service downtime.

During the entire process, communication between services must remain secure to ensure uninterrupted, encrypted traffic. This is essential for maintaining zero-trust security guarantees and ensuring operational reliability in production environments.

### As a user, it should not become harder to enable or manage mTLS in either single-zone or multi-zone deployments than it is today.

Currently, users can enable mTLS in both single-zone and multi-zone deployments just by configuring it in the Mesh resource. As we introduce features like trust domains, external identities, or federation, we must preserve this simplicity and avoid adding unnecessary complexity to the mTLS setup process.

### As a user, I want to be able to communicate with applications in the different trust domain.

If a user has a different trust domain in each zone, it should be possible to allow communication between these zones.

> ### As a Mesh operator, I want to be able to define a trustDomain for an entire mesh or zone

Author: Clarified

> It makes sense to allow users to define a trust domain for a specific part of the infrastructure: whether that's a zone or an entire mesh. However, this becomes tricky when balanced against the need for stricter security guarantees in other areas.

Author: It was a bit a conflict with `### As a Mesh operator, I should be prevented from configuring the same trust domain in multiple zones when not using SPIRE as the identity provider.` so I feel like removing this step and keeping `As a Mesh operator, I want to be able to define a trustDomain for an entire mesh or zone` makes more sense and it's clear

This point is a bit tricky for me, because on one hand, it makes sense to allow users to define a trust domain for an entire mesh (which may span multiple clusters). On the other hand, preventing users from using the same trust domain across different clusters improves security compliance.

By assigning unique trust domains per cluster, we achieve security isolation: if one cluster is compromised, the others remain secure. Additionally, it helps with identity scoping, since we can clearly determine the origin of each identity.

Default trust domain: Mesh name

### As a Mesh operator, I want to be able to migrate some workloads to another trust domain without interrupting traffic.

I have a workloads in a trust domain `example.com` and I want to move them to `example.org`. I should be able to do it and don't break the traffic.

### As a user, I want certificate to be valid for a short period of time

Based on the [specification](https://spiffe.io/docs/latest/spiffe-about/spiffe-concepts/#spiffe-workload-api) certificates should be short-lived.

## Out of scope

### As a user, I want to specify certificates for a specific outbound

> You should use MeshExternalService in this case to configure the outbound connection and provide the required certificates.

Author: Added note

### As a user, I want my dataplanes to communicate with Istio services without going through a gateway

In this scenario, you should also use MeshExternalService and include the necessary certificates in the configuration to enable direct mTLS communication.

### As a user, I want to use two different sources of CA (Kuma and SPIRE) at the same time

This is a complex use case and can make the configuration error prone and difficult to manage.
Instead, we recommend choosing one of the following approaches:

* Use Kuma as the CA and configure other zones or services to trust its certificates (i.e., share the same root/intermediate CA).

> * Use SPIRE as the CA and federate it across zones and meshes to provide a unified identity source.

Reviewer: However, I think it's ok here to say the Kuma-cp --> Spire migration must be made possible but won't be covered in this MADR.

> * Use SPIRE as the CA and federate it across zones and meshes to provide a unified identity source.

Author: Added section about migration and rephrased this one

# SPIFFE Compliance

* Status: accepted

Technical Story: https://github.com/kumahq/kuma/issues/13659

## Context and Problem Statement

In today’s service mesh systems, secure communication between services is essential. To make that possible, each service needs a clear and trusted identity: so we know exactly who is talking to whom.

One of the most trusted standards for defining these service identities is SPIFFE (Secure Production Identity Framework for Everyone). It gives us:

* A standard way to name services: `spiffe://<trust-domain>/<path>`
* A strict format for certificates, especially the Subject Alternative Name (SAN) - which must contain exactly one URI (the SPIFFE ID)
* Tools and rules for issuing, rotating, and checking these identities
* SPIFFE certificates identify workload, not an endpoint or API

> 

Reviewer 5: I don't fully understand this sentence here with the above context. Could you please explain some details?

Author: > SPIFFE certificates identify workload, not an endpoint or API

Each application might expose multiple endpoints. Currently, we create certificate for dataplane but we add identities of all endpoints to this certificate as SAN. That is not a good way, and certificate should just identify the workload and not each certificate for a separate API endpoint.

Right now, Kuma does something similar by generating SPIFFE-like IDs like `spiffe://<mesh-name>/<service-name>`. But it also adds extra SANs (like `kuma://kuma.io/version/v1`) based on tags. These extra SANs make Kuma's certificates non-compliant with SPIFFE, meaning they can’t be used directly with SPIFFE-based systems like SPIRE, or other tools that expect full compliance. Also, Kuma uses certificates to identify endpoints, whereas SPIFFE-based certificates are used to establish identity via SPIFFE IDs of the workload.

> In SPIFFE, a Trust Domain defines which services are allowed to trust each other. It’s like a boundary — services inside the same trust domain can safely communicate. To talk across trust domains (for example, between two companies or clusters), you need federation, which lets them trust each other's identities.

Reviewer 5: How about this?

```suggestion
In SPIFFE, a Trust Domain is a logical boundary for identity and trust, similar to a DNS domain. Services within the same trust domain can securely identify and trust each other. Communication across different trust domains (such as between organizations or clusters) requires federation, which establishes trust relationships so that identities from one trust domain are recognized and trusted by another.
```

Kuma currently uses the mesh name as the trust domain. While this works inside Kuma, it doesn’t clearly separate trust boundaries, and it limits how easily Kuma can connect with external services or other applications in the same trust domain.

This setup causes problems when:

* You want to connect Kuma with other SPIFFE-compliant systems outside the mesh
* You run multiple clusters that need to trust each other
* You work in environments that require strict identity standards for security or compliance reasons

To fully support SPIFFE and modern security practices, Kuma should:

* Issue certificates that include only one SPIFFE-compliant URI SAN (currently we generate `spiffe://<mesh>/<service>` for each inbound and `kuma://kuma.io/<tag_name>/<tag_value>` for each tag)
* Let users set the trust domain explicitly, not tie it to the mesh name

> * Allow using other systems (like SPIRE) to manage identities

Reviewer 5: > Let users set the trust domain explicitly, not tie it to the mesh name

So, it's a required property that the user must set, with no default value?

Author: We might introduce same default but it will be in a next PR

In today’s service mesh systems, secure communication between services is essential. To make that possible, each service needs a clear and trusted identity: so we know exactly who is talking to whom and who we can trust.

We should define two crucial concepts:
* **Identity** - (Who a workload is) — A workload's identity is the name encoded in its certificate, and this identity is considered valid only if the certificate is signed by a Trust.
* **Trust** - (Who to believe) - Trust defines which identities you accept as valid, and is established through trusted certificate authorities (CAs) that issue those identities. Trust is attached to trust domain, and there can be mutliple Trusts in the cluster.

One of the most trusted standards for defining these service identities is SPIFFE (Secure Production Identity Framework for Everyone). It gives us:

* A standard way to name services: `spiffe://<trust-domain>/<path>`
* A strict format for certificates, especially the Subject Alternative Name (SAN) - which must contain exactly one URI (the SPIFFE ID)
* Tools and rules for issuing, rotating, and checking these identities
* SPIFFE certificates identify workload, not an endpoint or API

> Right now, Kuma does something similar by generating SPIFFE-like IDs like `spiffe://<mesh-name>/<service-name>`. But it also adds extra SANs (like `kuma://kuma.io/version/v1`) based on tags. These extra SANs make Kuma's certificates non-compliant with SPIFFE, meaning they can’t be used directly with SPIFFE-based systems like SPIRE, or other tools that expect full compliance. Also, Kuma uses certificates to identify endpoints, whereas SPIFFE-based certificates are used to establish identity via SPIFFE IDs of the workload.

Reviewer 4: ```suggestion
Right now, Kuma does something similar by generating SPIFFE-like IDs like `spiffe://<mesh-name>/<service-name>`. But it also adds extra SANs (like `kuma://kuma.io/version/v1`) based on tags. These extra SANs make Kuma's certificates non-compliant with SPIFFE, meaning they can’t be used directly with SPIFFE-based systems like SPIRE, or other tools that expect full compliance. Also, Kuma uses certificates to identify endpoints (one Inbound), whereas SPIFFE-based certificates are used to establish identity via SPIFFE IDs of the workload (e.g. Deployment).
```

### As a user, I want to Trust from my zone to be synced to other zones automatically.

Trust could be automatically generated by the control plane and propagated to other zones. While this would improve automation and reduce configuration overhead, it is not critical at this stage and can be addressed in a future iteration.

### As a user, I want to use all available CA providers from legacy mTLS in new mTLS setup

Kuma currently supports various CA (Certificate Authority) backends for identity management. The new mTLS system should at least support the same set of CA providers, to ensure a smooth transition and preserve compatibility with existing deployments.

### As a user, I want to migrate from Kuma non-SPIFFE complient certificates to SPIRE

We should allow users to migrate from Kuma issued certificates, which are not SPIFFE-compliant, to using SPIRE. This may require an intermediate step to first migrate to SPIFFE-compliant certificates, but the details will be specified later.

### As a user, I want to automatically rotate the CA

> This could be straightforward to implement, but we shouldn't prioritize it at the moment.

Reviewer 4: I think it makes sense to add @lahabana comment here https://github.com/kumahq/kuma/pull/13699/files/7bf39d3fd8dd74da03cfbdf22f1c832d120f9fff..03927ffdadd7e6e30b897d0b7d6a8ef431502078#r2152062929

Author: It seems like SPIRE description, I would stay without specific description since it's not implementation but Context and Problem statement

## Out of scope

### As a user, I want to specify certificates for a specific outbound or inbound

You should use MeshExternalService in this case to configure the outbound connection and provide the required certificates. Dataplane has at the time only one Identity.

### As a user, I want my dataplanes to communicate with Istio services without going through a gateway

In this scenario, you should also use MeshExternalService and include the necessary certificates in the configuration to enable direct mTLS communication.

### As a user, I want to use two different sources of CA (Kuma non-SPIFFE compliant and SPIRE) at the same time

This is a complex use case and can make the configuration error prone and difficult to manage.
We want to add some limitation to avoid issues:

> * We should be add a Trust to SPIRE service while using Kuma certificate.

Reviewer 4: nitpick:

```suggestion
* We should be able to add a Trust to SPIRE service while using Kuma certificate.
```
